design/baremetal-operator/kubebuilder-migration.md

```diff
@@ -0,0 +1,202 @@
+  scaffolding for multiple API versions for the same resource type.
+
+### Non-Goals
+
+- Make any API changes to the BareMetalHost type.
+
+## Proposal
+
+As discussed in [the original v1alpha2 API
+proposal](https://github.com/metal3-io/metal3-docs/pull/101/), we will
+need to carry a modified version of kubebuilder to support float field
+types.  We can carry the fork either within the baremetal-operator
+repository or in a separate repository under the metal3-io org on
+github. It will be simpler to include it in the same source tree and
+have the Makefile there build the vendored tool.
+
+Kubebuilder initializes a project repository using a different layout
+than the old version of operator-sdk. It also uses a different
+approach to registering controllers, API types, etc. So, migrating
+will involve moving some existing code to new places and changing some
+of it. In particular:
+
+- kubebuilder places all packages at the top of the repository, not
+  under `pkg`
+- kubebuilder places all controllers in the same directory
+- main.go instantiates reconciler objects directly instead of calling
+  a factory, so some previously private parameters such as the
+  provisioner factory may need to be made public
+
+The [kubebuilder3
+branch](https://github.com/dhellmann/baremetal-operator/tree/kubebuilder3)
+of dhellmann's fork of the baremetal-operator repository is a
+proof-of-concept implementation of the migration. The host API is out
+of date, and the kubebuilder code probably is as well, but the same
+basic steps can be repeated when we are ready to do the work.
+
+As much effort as possible should be made to preserve the git history
+of files such as the API definition and the controller
+implementation. Unfortunately, these significant files will also have
+to be changed beyond their location within the repository, so
+preserving all of the history may not be possible.
+
+### User Stories
+
+#### Story 1
+
+As a contributor to the baremetal-operator, I want to use familiar
+tools and workflows so I can take advantage of my experience
+contributing elsewhere in the kubernetes community to work more
+efficiently.
+
+#### Story 2
+
+As a contributor to the baremetal-operator, I want to be able to make
+API changes to the BareMetalHost and support conversion web hooks to
+change between different storage versions.
+
+#### Story 3
+
+As a contributor to the baremetal-operator, I want to be able to use
+admission web hooks to validate BareMetalHost resources.
+
+## Design Details
+
+The process for creating the proof-of-concept migration was basically:
+
+1. Move everything in the git repository out of the way.
+2. Replace the copy of operator-sdk with a copy of kubebuilder, with a
+   similar patch to support float fields.
+3. Use kubebuilder init to create a new project in the repository,
+   setting the domain to an empty string.
+4. Use kubebuilder to create an API and controller for BareMetalHost,
```

Comment:
There's a migration guide for legacy Operator SDK projects: https://sdk.operatorframework.io/docs/building-operators/golang/project_migration_guide/

Reply:
Added to the references list at the end.